File: SocialBookAppInfrastructure/AuthorService.cs
using Microsoft.EntityFrameworkCore;
using SocialBookAppApplication;
using SocialBookAppDomain;

namespace SocialBookAppInfrastructure;

public class AuthorService : IAuthorService
{
    private readonly SocialBookAppContext _context;
    private readonly IRatingService _ratingService;

    public AuthorService(SocialBookAppContext context, IRatingService ratingService)
    {
        _context = context;
        _ratingService = ratingService;
    }

    public List<Author> GetAllAuthors()
    {
        return _context.Authors
            .Include(a => a.Books)
            .Include(a => a.Reviews)
            .ToList();
    }

    public Author GetAuthorById(int id)
    {
        var author = _context.Authors
            .Include(a => a.Books)
            .Include(a => a.Reviews)
            .FirstOrDefault(a => a.Id == id);
        if (author == null) throw new ArgumentException("Author not found.");
        return author;
    }

    public List<Book> GetAuthorBooks(int id)
    {
        var author = _context.Authors.FirstOrDefault(a => a.Id == id);
        if (author == null) throw new ArgumentException("Author not found.");
        return _context.Books.Where(b => b.Authors.Any(a => a.Id == id)).ToList();
    }

    public void AddAuthor(Author author)
    {
        _context.Authors.Add(author);
        _context.SaveChanges();
    }

    public void AssignBookToAuthor(int authorId, int bookId)
    {
        var author = _context.Authors
            .Include(a => a.Books)
            .FirstOrDefault(a => a.Id == authorId);
        if (author == null) throw new ArgumentException("Author not found.");

        var book = _context.Books.FirstOrDefault(b => b.Id == bookId);
        if (book == null) throw new ArgumentException("Book not found.");

        author.Books.Add(book);
        _context.SaveChanges();
    }

    public double GetRatingPercentage(List<Review> reviews)
    {
        return _ratingService.GetRatingPercentage(reviews);
    }
}

File: SocialBookAppVue/src/services/authorService.ts
import type { Author } from '../types/author.ts'
import { api } from './api'

export const getAuthors = async () => {
  const response = await api.get('/author')
  return response.data
}

export const getAuthorById = async (id: number) => {
  const response = await api.get(`/author/${id}`)
  return response.data
}

export const getAuthorBooks = async (id: number) => {
  const response = await api.get(`/author/${id}/books`)
  return response.data
}

export const getRatingPercentage = async (id: number) => {
  const response = await api.get(`/author/${id}/rating`)
  return response.data
}

export const addAuthor = async (author: Author) => {
  const response = await api.post('/author', author)
  return response.data
}

export const assignBook = async (authorId: number, bookId: number) => {
  const response = await api.put(`/author/${authorId}/book/${bookId}`)
  return response.data
}

export const unassignBook = async (authorId: number, bookId: number) => {
  await api.delete(`/author/${authorId}/book/${bookId}`)
}

export const updateAuthor = async (id: number, author: Author) => {
  const response = await api.put(`/author/${id}`, author)
  return response.data
}

export const deleteAuthor = async (id: number) => {
  await api.delete(`/author/${id}`)
}
